Clicking on external link returns error message (Android 13)

Sent by a user via email, let me know if info is missing:

Galaxy S20+5G, SM-G986B/DS, One UI-Version 5.1, Android 13;
 
Using Wikipedia_de_all_maxi_2023-01-09

"Geographic coordinates (of a location) cannot be copied from Kiwix and a window with apps to be selected (e.g. google Maps, OSMand, etc.) does not appear. This worked fine with Kiwix on my previous device (Samsung S10). Instead, the following message appears: "No apps could be installed for this data...." (the rest cannot be read even in landscape format, the message window does not increase in size).

![Screenshot_20230316_112437_Kiwix](https://user-images.githubusercontent.com/21675885/226312238-8a0d721f-688b-4dfe-ac34-5a4ab27a988b.jpg)

@kelson42 Yes! But in my case, it crashes. 
<details><Summary>Logs</summary> 

```

AndroidRuntime          org.kiwix.kiwixmobile                E  FATAL EXCEPTION: main
                                                                                                    Process: org.kiwix.kiwixmobile, PID: 20108
                                                                                                    java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:642)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1049)
                                                                                                    Caused by: java.lang.reflect.InvocationTargetException
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:632)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1049) 
                                                                                                    Caused by: java.net.MalformedURLException: unknown protocol: geo
                                                                                                    	at java.net.URL.<init>(URL.java:608)
                                                                                                    	at java.net.URL.<init>(URL.java:498)
                                                                                                    	at java.net.URL.<init>(URL.java:447)
                                                                                                    	at org.kiwix.kiwixmobile.core.utils.ExternalLinkOpener.requestOpenLink(ExternalLinkOpener.kt:64)
                                                                                                    	at org.kiwix.kiwixmobile.core.utils.ExternalLinkOpener.openExternalUrl(ExternalLinkOpener.kt:42)
                                                                                                    	at org.kiwix.kiwixmobile.core.main.CoreReaderFragment.openExternalUrl(CoreReaderFragment.kt:1203)
                                                                                                    	at org.kiwix.kiwixmobile.core.main.CoreWebViewClient.shouldOverrideUrlLoading(CoreWebViewClient.kt:73)

```

</details>

--- a/app/src/androidTest/java/org/kiwix/kiwixmobile/utils/ExternalLinkOpenerTest.kt
+++ b/app/src/androidTest/java/org/kiwix/kiwixmobile/utils/ExternalLinkOpenerTest.kt
@@ -65,6 +65,27 @@ internal class ExternalLinkOpenerTest {
   }
 
   @Test
+  internal fun alertDialogShowerOpensLinkIfGeoProtocolAdded() {
+    every { intent.resolveActivity(activity.packageManager) } returns mockk()
+    every { sharedPreferenceUtil.prefExternalLinkPopup } returns true
+    val uri = Uri.parse("geo:28.61388888888889,77.20833333333334")
+    every { intent.data } returns uri
+    val lambdaSlot = slot<() -> Unit>()
+    val externalLinkOpener = ExternalLinkOpener(activity, sharedPreferenceUtil, alertDialogShower)
+    externalLinkOpener.openExternalUrl(intent)
+    verify {
+      alertDialogShower.show(
+        KiwixDialog.ExternalLinkPopup,
+        capture(lambdaSlot),
+        any(),
+        any()
+      )
+    }
+    lambdaSlot.captured.invoke()
+    verify { activity.startActivity(intent) }
+  }
+
+  @Test
   internal fun alertDialogShowerDoesNoOpenLinkIfNegativeButtonIsClicked() {
     every { intent.resolveActivity(activity.packageManager) } returns mockk()
     every { sharedPreferenceUtil.prefExternalLinkPopup } returns true

--- a/core/src/main/java/org/kiwix/kiwixmobile/core/utils/ExternalLinkOpener.kt
+++ b/core/src/main/java/org/kiwix/kiwixmobile/core/utils/ExternalLinkOpener.kt
@@ -25,7 +25,6 @@ import org.kiwix.kiwixmobile.core.R
 import org.kiwix.kiwixmobile.core.extensions.toast
 import org.kiwix.kiwixmobile.core.utils.dialog.AlertDialogShower
 import org.kiwix.kiwixmobile.core.utils.dialog.KiwixDialog
-import java.net.URL
 import javax.inject.Inject
 
 class ExternalLinkOpener @Inject constructor(
@@ -61,7 +60,7 @@ class ExternalLinkOpener @Inject constructor(
         sharedPreferenceUtil.putPrefExternalLinkPopup(false)
         openLink(intent)
       },
-      url = URL(intent.data.toString())
+      uri = intent.data
     )
   }
 

--- a/core/src/main/java/org/kiwix/kiwixmobile/core/utils/dialog/AlertDialogShower.kt
+++ b/core/src/main/java/org/kiwix/kiwixmobile/core/utils/dialog/AlertDialogShower.kt
@@ -22,6 +22,7 @@ import android.app.Activity
 import android.app.Dialog
 import android.content.ClipData
 import android.content.ClipboardManager
+import android.net.Uri
 import android.text.Html
 import android.view.Gravity
 import android.view.ViewGroup.LayoutParams
@@ -32,7 +33,6 @@ import androidx.appcompat.app.AlertDialog
 import androidx.core.content.ContextCompat
 import org.kiwix.kiwixmobile.core.R
 import org.kiwix.kiwixmobile.core.extensions.getAttribute
-import java.net.URL
 import javax.inject.Inject
 
 class AlertDialogShower @Inject constructor(private val activity: Activity) :
@@ -44,10 +44,10 @@ class AlertDialogShower @Inject constructor(private val activity: Activity) :
     const val externalLinkBottomMargin = 0
   }
 
-  override fun show(dialog: KiwixDialog, vararg clickListeners: () -> Unit, url: URL?) =
-    create(dialog, *clickListeners, url = url).show()
+  override fun show(dialog: KiwixDialog, vararg clickListeners: () -> Unit, uri: Uri?) =
+    create(dialog, *clickListeners, uri = uri).show()
 
-  override fun create(dialog: KiwixDialog, vararg clickListeners: () -> Unit, url: URL?): Dialog {
+  override fun create(dialog: KiwixDialog, vararg clickListeners: () -> Unit, uri: Uri?): Dialog {
     return AlertDialog.Builder(activity)
       .apply {
         dialog.title?.let(this::setTitle)
@@ -70,7 +70,7 @@ class AlertDialogShower @Inject constructor(private val activity: Activity) :
               ?.invoke()
           }
         }
-        url?.let {
+        uri?.let {
           val frameLayout = FrameLayout(activity.baseContext)
 
           val textView = TextView(activity.baseContext).apply {
@@ -80,7 +80,7 @@ class AlertDialogShower @Inject constructor(private val activity: Activity) :
             setOnLongClickListener {
               val clipboard =
                 ContextCompat.getSystemService(activity.baseContext, ClipboardManager::class.java)
-              val clip = ClipData.newPlainText("External Url", "$url")
+              val clip = ClipData.newPlainText("External Url", "$uri")
               clipboard?.setPrimaryClip(clip)
               Toast.makeText(
                 activity.baseContext,
@@ -89,7 +89,7 @@ class AlertDialogShower @Inject constructor(private val activity: Activity) :
               ).show()
               true
             }
-            text = Html.fromHtml("</br><a href=$url> <b>$url</b>")
+            text = Html.fromHtml("</br><a href=$uri> <b>$uri</b>")
           }
           frameLayout.addView(textView)
           setView(frameLayout)

--- a/core/src/main/java/org/kiwix/kiwixmobile/core/utils/dialog/DialogShower.kt
+++ b/core/src/main/java/org/kiwix/kiwixmobile/core/utils/dialog/DialogShower.kt
@@ -18,18 +18,18 @@
 package org.kiwix.kiwixmobile.core.utils.dialog
 
 import android.app.Dialog
-import java.net.URL
+import android.net.Uri
 
 interface DialogShower {
   fun show(
     dialog: KiwixDialog,
     vararg clickListeners: (() -> Unit),
-    url: URL? = null
+    uri: Uri? = null
   )
 
   fun create(
     dialog: KiwixDialog,
     vararg clickListeners: (() -> Unit),
-    url: URL? = null
+    uri: Uri? = null
   ): Dialog
 }

